producer-c/producer-cloudwatch-integ/CanaryStreamCallbacks.cpp

```diff
@@ -0,0 +1,216 @@
+/**
+ * Kinesis Video Producer Continuous Retry Stream Callbacks
+ */
+#define LOG_CLASS "CanaryStreamCallbacks"
+#include "CanaryStreamCallbacks.h"
+
+STATUS createCanaryStreamCallbacks(Aws::CloudWatch::CloudWatchClient* cwClient,
+                                   PCHAR pStreamName,
+                                   PStreamCallbacks* ppStreamCallbacks, PCanaryStreamCallbacks pCanaryStreamCallbacks)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    Aws::CloudWatch::Model::Dimension dimension;
+
+    CHK(ppStreamCallbacks != NULL, STATUS_NULL_ARG);
+
+    // Allocate the entire structure
+    CHK(pCanaryStreamCallbacks != NULL, STATUS_NOT_ENOUGH_MEMORY);
+
+    // Set the version, self
+    pCanaryStreamCallbacks->streamCallbacks.version = STREAM_CALLBACKS_CURRENT_VERSION;
+    pCanaryStreamCallbacks->streamCallbacks.customData = (UINT64) pCanaryStreamCallbacks;
+    pCanaryStreamCallbacks->timeOfNextKeyFrame = new std::map<UINT64, UINT64>();
+    pCanaryStreamCallbacks->pStreamName = pStreamName;
+
+    pCanaryStreamCallbacks->pCwClient = cwClient;
+
+
+    dimension.SetName(pStreamName);
+    dimension.SetValue("ProducerSDK");
+
+    pCanaryStreamCallbacks->receivedAckDatum.SetMetricName("ReceivedAckLatency");
+    pCanaryStreamCallbacks->persistedAckDatum.SetMetricName("PersistedAckLatency");
+    pCanaryStreamCallbacks->streamErrorDatum.SetMetricName("StreamError");
+    pCanaryStreamCallbacks->currentFrameRateDatum.SetMetricName("FrameRate");
+    pCanaryStreamCallbacks->contentStoreAvailableSizeDatum.SetMetricName("StorageSizeAvailable");
+    pCanaryStreamCallbacks->currentViewDurationDatum.SetMetricName("CurrentViewDuration");
+    pCanaryStreamCallbacks->memoryAllocationSizeDatum.SetMetricName("MemoryAllocation");
+
+    pCanaryStreamCallbacks->receivedAckDatum.AddDimensions(dimension);
+    pCanaryStreamCallbacks->persistedAckDatum.AddDimensions(dimension);
+    pCanaryStreamCallbacks->streamErrorDatum.AddDimensions(dimension);
+    pCanaryStreamCallbacks->currentFrameRateDatum.AddDimensions(dimension);
+    pCanaryStreamCallbacks->contentStoreAvailableSizeDatum.AddDimensions(dimension);
+    pCanaryStreamCallbacks->currentViewDurationDatum.AddDimensions(dimension);
+    pCanaryStreamCallbacks->memoryAllocationSizeDatum.AddDimensions(dimension);
+    // Set callbacks
+    pCanaryStreamCallbacks->streamCallbacks.fragmentAckReceivedFn = CanaryStreamFragmentAckHandler;
+    pCanaryStreamCallbacks->streamCallbacks.streamErrorReportFn = CanaryStreamErrorReportHandler;
+    pCanaryStreamCallbacks->streamCallbacks.freeStreamCallbacksFn = CanaryStreamFreeHandler;
+
+CleanUp:
+
+    if (STATUS_FAILED(retStatus)) {
+        pCanaryStreamCallbacks = NULL;
```

Comment:
If you decide to allocate it in this fn then you need to free it too. Use same pattern as other providers in core c producer